cpp/ql/src/jsf/4.10 Classes/AV Rule 79.ql

```diff
@@ -159,6 +159,16 @@ predicate unreleasedResource(Resource r, Expr acquire, File f, int acquireLine)
     )
     and f = acquire.getFile()
     and acquireLine = acquire.getLocation().getStartLine()
+
+    // check that any destructor for this class has a block; if it doesn't,
+    // we must be missing information.
+    and forall(Destructor d |
+      d = r.getDeclaringType().getAMember() and
+      not d.isCompilerGenerated() and
+      not d.isDefaulted() and
+      not d.isDeleted() |
+      exists(d.getBlock())
+    ) 
```

Comment:
@ian-semmle How does this change interact with your upcoming changes to reduce db sizes? Do I remember correctly that you're changing it so unused members of templates will not have their signatures extracted at all?

Reply:
In master,
```
template <typename T>
class Cl {
public:
    ~Cl();
};

void fun(Cl<int> x) {
}
```
has Functions
```
Cl<T>::~Cl() -> void
Cl<int>::~Cl() -> void
```
but in discover_walk it has only
```
Cl<T>::~Cl() -> void
```

Reply:
So perhaps for template instantiations we will also have to accept if the corresponding template has a destructor with a block???

@ian-semmle, is it reasonable for me to test these changes on `discover_walk` (i.e. is the branch ready enough)?

Reply:
Yes; if there are any problems with it then I'd like to hear about them!

Reply:
IIUC, with the `discover_walk` change this query will believe there's no destructor on the instantiation whereas without `discover_walk` it would see a destructor declaration without a body and assume that the body did the right thing.

I agree this means the query will have to check the corresponding template to see if there's a constructor there. That change could be done on the `next` branch since it would fix a problem that won't be observed on `master`, or it could be done on `master` to reduce potential conflicts in the future. In any case, we should make sure this problem is tested for on `master`.

Reply:
IIUC, with the `discover_walk` change this query will believe there's no destructor on the instantiation whereas without `discover_walk` it would see a destructor declaration without a body and assume that the body did the right thing.

I agree this means the query will have to check the corresponding template to see if there's a constructor there. That change could be done on the `next` branch since it would fix a problem that won't be observed on `master`, or it could be done on `master` to reduce potential conflicts in the future. In any case, we should make sure this problem is tested for on `master`.

Reply:
IIUC, with the `discover_walk` change this query will believe there's no destructor on the instantiation whereas without `discover_walk` it would see a destructor declaration without a body and assume that the body did the right thing.

I agree this means the query will have to check the corresponding template to see if there's a constructor there. That change could be done on the `next` branch since it would fix a problem that won't be observed on `master`, or it could be done on `master` to reduce potential conflicts in the future. In any case, we should make sure this problem is tested for on `master`.

Reply:
I've just rebased and added a fix (two commits) for this, which passes the tests with or without `discover_walk`.